curp/src/server/curp_node.rs

```diff
@@ -62,22 +61,91 @@ pub(super) enum CurpError {
     /// Io error
     #[error("io error, {0}")]
     IO(#[from] io::Error),
-    /// Internal error
+    /// Currently, Internal Error includes the following three types of errors:
+    /// 1. failed to allocate a new snapshot
+    /// 2. failed to reset the command executor by snapshot
+    /// 3. failed to get last applied index from command executor.
     #[error("internal error, {0}")]
     Internal(String),
+    /// Curp Server is shutting down
+    #[error("cluster shutdown")]
+    ShuttingDown,
+    /// If client sent a wait synced request to a non-leader
+    #[error("redirect to {0:?}, term {1}")]
+    Redirect(Option<ServerId>, u64),
 }
 
 impl From<bincode::Error> for CurpError {
     fn from(err: bincode::Error) -> Self {
         Self::EncodeDecode(err.to_string())
     }
 }
+
 impl From<PbSerializeError> for CurpError {
     fn from(err: PbSerializeError) -> Self {
         Self::EncodeDecode(err.to_string())
     }
 }
 
+/// set `error-label` in `tonic::metadata::MetadataMap`
+fn set_error_label_for_status(status: &mut tonic::Status, label: &str) {
```

Comment:
`Metadata`(Also could be considered as `http::HeaderMap`) does not seem to be a suitable payload for error messages. We need to write some code to parse the string and take some actions. I think we can serialize error-related objects into JSON and place them in the "details" field, which can maintain the hierarchy of errors.

Reply:
Done